In [1]:
import kagglehub
import os
import pandas as pd
import numpy as np

# Download latest version
path = kagglehub.dataset_download("dgomonov/new-york-city-airbnb-open-data")

print("Path to dataset files:", path)

if os.path.isdir(path):
  print(True)

contents = os.listdir(path)
contents

mydataset = path + "/" + contents[0]
mydataset


df = pd.read_csv(mydataset)


Using Colab cache for faster access to the 'new-york-city-airbnb-open-data' dataset.
Path to dataset files: /kaggle/input/new-york-city-airbnb-open-data
True


### A. Discovering



Display the first 10 rows

In [2]:
df.head(10)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   
5  5099         Large Cozy 1 BR Apartment In Midtown East     7322   
6  5121                                   BlissArtsSpace!     7356   
7  5178                  Large Furnished Room Near B'way      8967   
8  5203                Cozy Clean Guest Room - Family Apt     7490   
9  5238                Cute & Cozy Lower East Side 1 bdrm     7549   

     host_name neighbourhood_group       neighbourhood  latitude  longitude  \
0         John            Brooklyn          Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan             Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan              Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn        Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan         East Harlem  40.79851  -73.94399   
5        Chris           Manhattan         Murray Hill  40.74767  -73.97500   
6        Garon            Brooklyn  Bedford-Stuyvesant  40.68688  -73.95596   
7     Shunichi           Manhattan      Hell's Kitchen  40.76489  -73.98493   
8    MaryEllen           Manhattan     Upper West Side  40.80178  -73.96723   
9          Ben           Manhattan           Chinatown  40.71344  -73.99037   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   
5  Entire home/apt    200               3                 74  2019-06-22   
6     Private room     60              45                 49  2017-10-05   
7     Private room     79               2                430  2019-06-24   
8     Private room     79               2                118  2017-07-21   
9  Entire home/apt    150               1                160  2019-06-09   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0  
5               0.59                               1               129  
6               0.40                               1                 0  
7               3.47                               1               220  
8               0.99                               1                 0  
9               1.33                               4               188

Display the column names:


In [3]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

Display the dataframe shape:

In [4]:
df.shape

(48895, 16)

Identify the numerical features:

In [5]:
numerical_features = df.select_dtypes(include=np.number).columns
print("Numerical features:", list(numerical_features))

Numerical features: ['id', 'host_id', 'latitude', 'longitude', 'price', 'minimum_nights', 'number_of_reviews', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365']


Identify the categorical features

In [6]:
categorical_features = df.select_dtypes(include=object).columns
print("Categorical features:", list(categorical_features))

Categorical features: ['name', 'host_name', 'neighbourhood_group', 'neighbourhood', 'room_type', 'last_review']


What are the top 5 questions this dataset might help answer?

1. Price Analysis: What is the average price for different room types across various neighbourhood groups in NYC?

2. Popularity & Reviews: Is there a correlation between the price and the number of reviews or reviews per month?

3. Geographic Trends: Do prices vary geographically and Which neigbourhoods have the highest and lowest concentration of listings?

4. Host Activity: Do high-volume hosts (high calculated_host_listings_count) dominate certain neighborhoods or price points?

5. Availability & Demand: How does the minimum nights requirement impact listing availability throughout the year?


## B. Structuring

Check if the dataset has proper column types

In [7]:
print(df.dtypes)

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object


The dataset seems to have correct object types except last_review column which can be benefecial to convert from object to date-time

Check if the dataset has index issues


In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48884 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48884 non-null  int64         
 1   name                            48884 non-null  object        
 2   host_id                         48884 non-null  int64         
 3   host_name                       48884 non-null  object        
 4   neighbourhood_group             48884 non-null  object        
 5   neighbourhood                   48884 non-null  object        
 6   latitude                        48884 non-null  float64       
 7   longitude                       48884 non-null  float64       
 8   room_type                       48884 non-null  object        
 9   price                           48884 non-null  int64         
 10  minimum_nights                  48884 non-null  int64         
 11  number_

According to the df.info() output, the dataset uses a default RangeIndex (from 0 to 48894), which means there are no immediate index issues.

Check if the dataset has duplicated rows

In [9]:
print(f"Number of duplicated rows: {df.duplicated().sum()}")

Number of duplicated rows: 0


Convert columns to correct types (e.g., dates → datetime)

In [10]:
df['last_review'] = pd.to_datetime(df['last_review'])
print(f'last_review column changed to {df["last_review"].dtype}')

last_review column changed to datetime64[ns]


Identify at least one structural issue and describe how to fix it.

One structural issue is the presence of missing values in the reviews_per_month, name, and host_name columns. Specifically, reviews_per_month often contains NaN for listings with zero reviews, which should logically be 0. Missing values in name and host_name can also be handled.

## C. Cleaning

Check for missing values

In [11]:
print(df.isna().sum())

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64


Check for outliers in minimum_nights

In [19]:
print('Descriptive statistics for minimum_nights:')
print(df['minimum_nights'].describe())
from scipy.stats import zscore

# Calculate Z-scores for 'minimum_nights'
z_scores = np.abs(zscore(df['minimum_nights']))

# Define a threshold for extreme outliers (e.g., Z-score > 3)
# A Z-score greater than 3 (or less than -3) is often considered an extreme outlier.
outlier_threshold = 3

# Identify outliers
outliers_from_mean = df[z_scores > outlier_threshold]

print(f"Number of outliers in 'minimum_nights' very far from the mean (Z-score > {outlier_threshold}): {len(outliers_from_mean)}")
print("These outliers represent values that are more than 3 standard deviations away from the mean.")

Descriptive statistics for minimum_nights:
count    48895.000000
mean         7.029962
std         20.510550
min          1.000000
25%          1.000000
50%          3.000000
75%          5.000000
max       1250.000000
Name: minimum_nights, dtype: float64
Number of outliers in 'minimum_nights' very far from the mean (Z-score > 3): 327
These outliers represent values that are more than 3 standard deviations away from the mean.


Check for invalid values (e.g., price ≤ 0)

In [20]:
invalid_prices = df[df['price'] <= 0]
print(f"Number of listings with price <= 0: {len(invalid_prices)}")
if not invalid_prices.empty:
    print("Sample of listings with invalid prices:")
    print(invalid_prices[['id', 'name', 'price']].head())

Number of listings with price <= 0: 11
Sample of listings with invalid prices:
             id                                               name  price
23161  18750597  Huge Brooklyn Brownstone Living, Close to it all.      0
25433  20333471      ★Hostel Style Room | Ideal Traveling Buddies★      0
25634  20523843    MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)      0
25753  20608117                    Sunny, Quiet Room in Greenpoint      0
25778  20624541      Modern apartment in the heart of Williamsburg      0


Apply a missing-value strategy (drop, fill, or flag)

In [23]:
# 1. Handle missing values in 'reviews_per_month': fill with 0
df['reviews_per_month'] = df['reviews_per_month'].fillna(0)
print(f"NaNs in 'reviews_per_month' after filling: {df['reviews_per_month'].isnull().sum()}")

# 2. Handle missing values in 'name' and 'host_name': fill with 'Unknown'
df['name'] = df['name'].fillna('Unknown')
df['host_name'] = df['host_name'].fillna('Unknown')
print(f"NaNs in 'name' after filling: {df['name'].isnull().sum()}")
print(f"NaNs in 'host_name' after filling: {df['host_name'].isnull().sum()}")

# 'last_review' NaN values are consistent with 'reviews_per_month' being 0, so no specific fill is needed unless a default date is desired.

# 3. Handle invalid prices (price <= 0): drop these rows
initial_rows = df.shape[0]
df = df[df['price'] > 0]
removed_rows = initial_rows - df.shape[0]
print(f"Removed {removed_rows} rows with invalid prices (price <= 0).")

print("\nMissing values after applying strategies:")
print(df.isnull().sum())

NaNs in 'reviews_per_month' after filling: 0
NaNs in 'name' after filling: 0
NaNs in 'host_name' after filling: 0
Removed 0 rows with invalid prices (price <= 0).

Missing values after applying strategies:
id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10051
reviews_per_month                     0
calculated_host_listings_count        0
availability_365                      0
dtype: int64


Apply an outlier-handling strategy

In [24]:
percentile_99_5 = df['minimum_nights'].quantile(0.995)
df['minimum_nights'] = np.where(df['minimum_nights'] > percentile_99_5, percentile_99_5, df['minimum_nights'])

print(f"'minimum_nights' values have been capped at the 99.5th percentile ({percentile_99_5}).")
print("Descriptive statistics for 'minimum_nights' after capping:")
print(df['minimum_nights'].describe())

'minimum_nights' values have been capped at the 99.5th percentile (90.0).
Descriptive statistics for 'minimum_nights' after capping:
count    48884.000000
mean         6.450127
std         11.176115
min          1.000000
25%          1.000000
50%          3.000000
75%          5.000000
max         90.000000
Name: minimum_nights, dtype: float64


## D. Enriching

Create at least 3 new variables:

price_per_minimum_night

review_rate_per_month (combine ratings + frequency)

is_expensive (boolean flag based on price threshold)

In [26]:
df['price_per_minimum_night'] = df['price'] / df['minimum_nights']

# 2. review_rate_score (combine ratings + frequency)
# This uses a simple multiplicative score: Total reviews * (1 + frequency)
df['review_rate_score'] = df['number_of_reviews'] * (1 + df['reviews_per_month'])

# 3. is_expensive (boolean flag based on price threshold - using the 75th percentile)
price_threshold = df['price'].quantile(0.75)
df['is_expensive'] = df['price'] > price_threshold

print(df[['price', 'minimum_nights', 'price_per_minimum_night', 'review_rate_score', 'is_expensive']].head())

   price  minimum_nights  price_per_minimum_night  review_rate_score  \
0    149             1.0                    149.0              10.89   
1    225             1.0                    225.0              62.10   
2    150             3.0                     50.0               0.00   
3     89             1.0                     89.0            1522.80   
4     80            10.0                      8.0               9.90   

   is_expensive  
0         False  
1          True  
2         False  
3         False  
4         False  


Create 1 aggregated summary table

(e.g. average price per neighborhood, number of hosts per borough)

In [27]:

# put your answer here

# Aggregation: Average price and count per Neighbourhood Group
neighbourhood_summary = df.groupby('neighbourhood_group').agg(
    average_price=('price', 'mean'),
    total_listings=('id', 'count'),
    median_minimum_nights=('minimum_nights', 'median')
).sort_values(by='average_price', ascending=False).reset_index()

print(neighbourhood_summary)

  neighbourhood_group  average_price  total_listings  median_minimum_nights
0           Manhattan     196.884903           21660                    3.0
1            Brooklyn     124.438915           20095                    3.0
2       Staten Island     114.812332             373                    2.0
3              Queens      99.517649            5666                    2.0
4               Bronx      87.577064            1090                    2.0


## E. Validating

Create at least 3 validation checks

Price must be positive.

Host ID should not be null.

Minimum nights must be ≥ 1.

Generate a validation report:

Count of rows violating each rule.

Decide if invalid rows (if any) should be removed or corrected?

In [28]:
# put your answer here

# Validation Check 1: Price must be positive.
rule_1_violations = df[df['price'] <= 0].shape[0]

# Validation Check 2: Host ID should not be null.
rule_2_violations = df['host_id'].isnull().sum()

# Validation Check 3: Minimum nights must be >= 1.
rule_3_violations = df[df['minimum_nights'] < 1].shape[0]

# Generate a validation report:
validation_report = pd.DataFrame({
    'Validation Rule': ['Price must be positive', 'Host ID should not be null', 'Minimum nights must be >= 1'],
    'Violating Rows Count': [rule_1_violations, rule_2_violations, rule_3_violations]
})

print("Validation Report:")
print(validation_report)

"""
Decide if invalid rows (if any) should be removed or corrected?

Decision: The invalid rows were already REMOVED during the cleaning stage (C). The validation report shows 0 violations for these core business rules, confirming the success of the cleaning step. For critical issues (invalid price, extreme nights), removal is the preferred strategy.
"""

Validation Report:
               Validation Rule  Violating Rows Count
0       Price must be positive                     0
1   Host ID should not be null                     0
2  Minimum nights must be >= 1                     0


'\nDecide if invalid rows (if any) should be removed or corrected?\n\nDecision: The invalid rows were already REMOVED during the cleaning stage (C). The validation report shows 0 violations for these core business rules, confirming the success of the cleaning step. For critical issues (invalid price, extreme nights), removal is the preferred strategy.\n'

## F. Publishing

Select final columns for publication.

In [29]:
# put your answer here

final_columns = [
    'host_id',
    'neighbourhood_group',
    'neighbourhood',
    'latitude',
    'longitude',
    'room_type',
    'price',
    'minimum_nights',
    'number_of_reviews',
    'reviews_per_month',
    'calculated_host_listings_count',
    'availability_365',
    'price_per_minimum_night',
    'review_rate_score',
    'is_expensive'
]

df_cleaned = df[final_columns]
print("Final columns selected.")

Final columns selected.


Export cleaned dataset as "cleaned_airbnb.csv"

In [30]:

df_cleaned.to_csv("cleaned_airbnb.csv", index=False)
print('Exported cleaned dataset as "cleaned_airbnb.csv" [fileTag: cleaned_airbnb.csv]')

Exported cleaned dataset as "cleaned_airbnb.csv" [fileTag: cleaned_airbnb.csv]


Export summary reports (e.g., validation or aggregations) as CSV.

In [31]:
validation_report.to_csv("validation_report.csv", index=False)
neighbourhood_summary.to_csv("neighbourhood_summary.csv", index=False)
print('Exported summary reports: "validation_report.csv" [fileTag: validation_report.csv] and "neighbourhood_summary.csv" [fileTag: neighbourhood_summary.csv]')


Exported summary reports: "validation_report.csv" [fileTag: validation_report.csv] and "neighbourhood_summary.csv" [fileTag: neighbourhood_summary.csv]


In [33]:
# 1. Calculate the 75th percentile (Q3) for price, grouped by 'room_type'.
# This creates a unique 'local' threshold for each room type.
room_type_q3 = df.groupby('room_type')['price'].quantile(0.75).reset_index(name='Q3_threshold')

# 2. Merge the thresholds back into the main DataFrame.
df_merged = pd.merge(df, room_type_q3, on='room_type', how='left')

# 3. Create the conditional flag: 'is_expensive_by_roomtype'.
# A listing is expensive IF its price is GREATER THAN its specific room type's Q3.
df_merged['is_expensive_by_roomtype'] = df_merged['price'] > df_merged['Q3_threshold']

# 4. Print the thresholds (Proof that price changes by type)
print("--- Q3 Price Thresholds by Room Type (The Benchmark) ---")
print(room_type_q3.to_markdown(index=False))

# 5. Validate/Prove the 25% claim for the 'Entire home/apt' category
entire_home_data = df_merged[df_merged['room_type'] == 'Entire home/apt']
entire_home_expensive_count = entire_home_data['is_expensive_by_roomtype'].sum()
entire_home_total = entire_home_data.shape[0]
entire_home_percentage = (entire_home_expensive_count / entire_home_total) * 100

print("\n--- Validation of Conditional Flag (Proof of Methodology) ---")
print(f"Total 'Entire home/apt' Listings: {entire_home_total}")
print(f"Percentage Flagged as Expensive (Relative to Entire Homes): {entire_home_percentage:.2f}%")

--- Q3 Price Thresholds by Room Type (The Benchmark) ---
| room_type       |   Q3_threshold |
|:----------------|---------------:|
| Entire home/apt |            229 |
| Private room    |             95 |
| Shared room     |             75 |

--- Validation of Conditional Flag (Proof of Methodology) ---
Total 'Entire home/apt' Listings: 25407
Percentage Flagged as Expensive (Relative to Entire Homes): 24.88%
